contracts/NftCustom.sol

```diff
@@ -0,0 +1,205 @@
+    public
+    view
+    override(ERC721, AccessControl)
+    returns (bool)
+    {
+        return ERC721.supportsInterface(interfaceId);
+    }
+
+    function owner() public view returns (address) {
+        return _owner;
+    }
+
+    function _baseURI() 
+    internal
+    view 
+    virtual 
+    override(ERC721)
+    returns (string memory) {
+        return baseURI;
+    }
+
+
+    function updateTokenUri(uint256 _tokenId, string memory _tokenUri, bool _isFreezeTokenUri)
+    public
+    onlyRole(MINTER_ROLE) {
+        require(_exists(_tokenId), "NFT: update URI query for nonexistent token");
+        require(isFreezeTokenUris == false, "NFT: Token uris are frozen globally");
+        require(freezeTokenUris[_tokenId] != true, "NFT: Token is frozen");
+        require(_isFreezeTokenUri || (bytes(_tokenUri).length != 0), "NFT: Either _tokenUri or _isFreezeTokenUri=true required");
+
+        if (bytes(_tokenUri).length != 0) {
+            require(keccak256(bytes(tokenURI(_tokenId))) != keccak256(bytes(string(abi.encodePacked(_baseURI(), _tokenUri)))), "NFT: New token URI is same as updated");
+            _setTokenURI(_tokenId, _tokenUri);
+        }
+        if (_isFreezeTokenUri) {
+            freezeTokenUris[_tokenId] = true;
+            emit PermanentURI(tokenURI(_tokenId), _tokenId);
+        }
+    }
+
+    function burn(uint256 _tokenId)
+    public
+    onlyRole(MINTER_ROLE) {
+        require(_exists(_tokenId), "Burn for nonexistent token");
+        _burn(_tokenId);
+    }
+
+    function update(string memory _newBaseURI, bool _freezeAllTokenUris)
+    public
+    onlyRole(MINTER_ROLE) {
+        require(isFreezeTokenUris == false, "NFT: Token uris are already frozen");
+        baseURI = _newBaseURI;
+        if (_freezeAllTokenUris) {
+            freezeAllTokenUris();
+        }
+    }
+
+    function freezeAllTokenUris()
+    public
+    onlyRole(MINTER_ROLE) {
+        require(isFreezeTokenUris == false, "NFT: Token uris are already frozen");
+        isFreezeTokenUris = true;
+
+        emit PermanentURIGlobal();
+    }
+
+    function totalSupply() public view virtual returns (uint256) {
+        return _allTokens.length;
+    }
+
+    function tokenOfOwnerByIndex(address owner, uint256 index) public view virtual returns (uint256) {
+        require(index < balanceOf(owner), "ERC721: owner index out of bounds");
+        return _ownedTokens[owner][index];
+    }
+
+    function tokenByIndex(uint256 index) public view virtual returns (uint256) {
+        require(index < totalSupply(), "ERC721: global index out of bounds");
+        return _allTokens[index];
+    }
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal virtual override {
+        super._beforeTokenTransfer(from, to, tokenId);
+
+        if (from == address(0)) {
+            _addTokenToAllTokensEnumeration(tokenId);
+        } else if (from != to) {
```

Comment:
Can you explain this check please? What does `from != to` mean? And also later `to != from`?

Reply:
Also, is this functionality covered by unit tests?

Reply:
This code was taken from ERC721Enumerable
It updates index-related mappings/arrays while minting/transferring/burning tokens.
Unit tests are covering higher-level public funcs (totalSupply/tokenOfOwnerByIndex/tokenByIndex before mint/after mint/after burn), including negative cases for out-of-bound indexes.